pkg/client/client.go

```diff
@@ -352,6 +387,188 @@ func (c *serviceImpl) doRequest(method, path string, reqpl, respl any) error {
+	infos, err := u.getFilesInfos()
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	jobsChan := make(chan uploadChunkJob)
+	resultsChan := u.startWorkers(ctx, jobsChan)
+	go func() {
+		defer close(jobsChan)
+		for _, info := range infos {
+			for i := 0; i < info.TotalChunks; i++ {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					j := uploadChunkJob{
+						Filename:       info.Name,
+						ChunkNumber:    i + 1,
+						TotalChunks:    info.TotalChunks,
+						ChunkSizeBytes: u.ChunkSizeBytes,
+					}
+					jobsChan <- j
+				}
+			}
+		}
+	}()
+	// Only first error will be returned.
+	var returnErr error
+	for err := range resultsChan {
+		if returnErr != nil {
+			continue
+		}
+		if err != nil {
+			returnErr = err
+			cancel()
+			// Do not return from here and let the cancellation logic to propagate, resultsChan
+			// will be closed eventually.
+		}
+	}
+	return returnErr
+}
+
+func (u *filesUploader) getFilesInfos() ([]fileInfo, error) {
+	var infos []fileInfo
+	for _, name := range u.Filenames {
+		stat, err := os.Stat(name)
+		if err != nil {
+			return nil, err
+		}
+		info := fileInfo{
+			Name:        name,
+			TotalChunks: int((stat.Size() + u.ChunkSizeBytes - 1) / u.ChunkSizeBytes),
+		}
+		infos = append(infos, info)
+	}
+	return infos, nil
+}
+
+func (u *filesUploader) startWorkers(ctx context.Context, jobsChan <-chan uploadChunkJob) <-chan error {
+	agg := make(chan error)
+	wg := sync.WaitGroup{}
+	for i := 0; i < openConnections; i++ {
+		wg.Add(1)
+		w := uploadChunkWorker{
+			Client:      u.Client,
+			EndpointURL: u.EndpointURL,
+			DumpOut:     u.DumpOut,
+			JobsChan:    jobsChan,
+		}
+		go func(ch <-chan error) {
+			for err := range ch {
+				agg <- err
+			}
+			wg.Done()
+		}(w.Start())
+	}
+	go func() {
+		wg.Wait()
+		close(agg)
+	}()
+	return agg
+}
+
+type uploadChunkJob struct {
+	Filename string
+	// A number between 1 and `TotalChunks`. The n-th chunk represents a segment of data within the file with size
+	// `ChunkSizeBytes` starting the `(n-1) * ChunkSizeBytes`-th byte.
+	ChunkNumber    int
+	TotalChunks    int
+	ChunkSizeBytes int64
+}
+
+type uploadChunkWorker struct {
+	Client      *http.Client
+	EndpointURL string
+	DumpOut     io.Writer
+	JobsChan    <-chan uploadChunkJob
+
+	bodyBuff bytes.Buffer
+}
+
+// Returns a channel that will return the result for each of the handled `uploadChunkJob` instances.
+func (w *uploadChunkWorker) Start() <-chan error {
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		for job := range w.JobsChan {
+			ch <- w.upload(job)
+		}
+	}()
+	return ch
+}
+
+func (w *uploadChunkWorker) upload(job uploadChunkJob) error {
+	file, err := os.Open(job.Filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if _, err := file.Seek(int64(job.ChunkNumber-1)*job.ChunkSizeBytes, 0); err != nil {
+		return err
+	}
+	writer := multipart.NewWriter(&w.bodyBuff)
```

Comment:
This code reads the entire chunk into memory before sending it to the server. For smaller chunks this isn't a problem, but nothing guarantees the chunks will always be small, especially with big files such as `super.img`.
If you replace the in-memory buffer `w.bodyBuff` with a [pipe](https://pkg.go.dev/io#Pipe) that problem goes away at the cost of having to write the request body in an additional routine, but luckily you're already using a context to cancel the request if there are errors on that routine. Double check me on this, but if you cancel the request through the context `w.Client.Do(req)` will return an error which you can return so there aren't many changes required to make this work.

Reply:
great point, used pipe instead as suggested. done